File: java/config/config-store/src/main/java/org/nhindirect/config/store/Anchor.java
package org.nhindirect.config.store;
/* 
Copyright (c) 2010, NHIN Direct Project
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
   in the documentation and/or other materials provided with the distribution.  
3. Neither the name of the The NHIN Direct Project (nhindirect.org) nor the names of its contributors may be used to endorse or promote 
   products derived from this software without specific prior written permission.
   
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE 
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
THE POSSIBILITY OF SUCH DAMAGE.
*/

import java.io.ByteArrayInputStream;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.sql.Timestamp;
import java.util.Calendar;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.Lob;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;


@Entity
@Table(name="anchor")
/**
 * The JPA Domain class
 */
public class Anchor {
	
	private String owner;
	private String thumbprint;
	private long   certificateId;
	private byte[] certificateData;
	private long   id; 
	private Calendar createTime;
	private Calendar validStartDate;
	private Calendar validEndDate;
	private EntityStatus status;
	private boolean incoming;
	private boolean outgoing;
	
	@Column(name="owner")
	public String getOwner() {
		return owner;
	}
	public void setOwner(String owner) {
		this.owner = owner;
	}
	
	@Column(name="thumbprint")
	public String getThumbprint() {
		return thumbprint;
	}
	
	public void setThumbprint(String thumbprint) {
		this.thumbprint = thumbprint;
	}
	
	@Column(name="certificateData")
	@Lob
	public byte[] getData() {
		return certificateData;
	}

	public void setData(byte[] data) throws CertificateException {
		certificateData  = data;
		if (data == Certificate.NULL_CERT) {
			setThumbprint("");
		}
		else {
			loadCertFromData();
		}
	}
	
	
	@Id
	@Column(name="id",nullable=false)
	@GeneratedValue(strategy=GenerationType.SEQUENCE)
	public long getId() {
		return id;
	}
	
	public void setId(long id) {
		this.id = id;
	}
	
	@Temporal(TemporalType.TIMESTAMP)
	public Calendar getCreateTime() {
		return createTime;
	}

	public void setCreateTime(Calendar timestamp) {
		createTime = timestamp;
	}
	
	@Temporal(TemporalType.TIMESTAMP)
	public Calendar getValidStartDate() {
		return validStartDate;
	}
	
	public void setValidStartDate(Calendar validStartDate) {
		this.validStartDate = validStartDate;
	}
	
	@Temporal(TemporalType.TIMESTAMP)
	public Calendar getValidEndDate() {
		return validEndDate;
	}
	
	public void setValidEndDate(Calendar validEndDate) {
		this.validEndDate = validEndDate;
	}
	
	@Enumerated
	public EntityStatus getStatus() {
		return status;
	}
	
	public void setStatus(EntityStatus status) {
		this.status = status;
	}
	
	@Column(name="forIncoming")
	public boolean isIncoming() {
		return incoming;
	}
	
	public void setIncoming(boolean incoming) {
		this.incoming = incoming;
	}
	
	@Column(name="forOutgoing")
	public boolean isOutgoing() {
		return outgoing;
	}
	
	public void setOutgoing(boolean outgoing) {
		this.outgoing = outgoing;
	}
	
	private void loadCertFromData() throws CertificateException {
		X509Certificate cert = null;
		try {
			validate();
			ByteArrayInputStream bais = new ByteArrayInputStream(certificateData);
			cert = (X509Certificate)CertificateFactory.getInstance("X.509").generateCertificate(bais);
			setThumbprint(Thumbprint.toThumbprint(cert).toString());
			bais.close();
		}
		catch (Exception e) {
			setData(Certificate.NULL_CERT);
			throw new CertificateException("Data cannot be converted to a valid X.509 Certificate", e);
		}
	}
	
    private boolean hasData() {
    	return ((certificateData != null) && (!certificateData.equals(Certificate.NULL_CERT))) ? true : false;
    }
    
	
	public void validate() throws CertificateException {
		if (!hasData()) {
			throw new CertificateException("Invalid Certificate: no certificate data exists");
		}
	}

}

File: csharp/config/config.store.sqlserver/Entity/Anchor.cs
/* 
 Copyright (c) 2010, Direct Project
 All rights reserved.

 Authors:
    Joe Shook       Joseph.Shook@Surescripts.com

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of The Direct Project (directproject.org) nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
*/


using System.Security.Cryptography.X509Certificates;

namespace Health.Direct.Config.Store.Entity
{
    public class Anchor
    {
        public const int MaxOwnerLength = 400;
        
        string m_owner;
        byte[] m_data;
        
        public Anchor()
        {
        }

        public Anchor(string owner, byte[] sourceFileBytes, string password)
            : this(owner, Certificate.Import(sourceFileBytes, password))
        {
        }
        
        
        public Anchor(string owner, X509Certificate2 certificate)
            : this(owner, certificate, true, true)
        {
        }
        
        public Anchor(string owner, X509Certificate2 certificate, bool forIncoming, bool forOutgoing)
        {
            if (certificate == null)
            {
                throw new ConfigStoreException(ConfigStoreError.InvalidCertificate);
            }
            
            this.Owner = owner;
            this.Thumbprint = certificate.Thumbprint;
            this.Data = certificate.RawData;
            this.CreateDate = DateTimeHelper.Now;
            this.ValidStartDate = certificate.NotBefore;
            this.ValidEndDate = certificate.NotAfter;
            this.ForIncoming = forIncoming;
            this.ForOutgoing = forOutgoing;
        }

        public string Owner
        {
            get
            {
                return m_owner;
            }
            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ConfigStoreException(ConfigStoreError.InvalidOwnerName);
                }
                
                if (value.Length > MaxOwnerLength)
                {
                    throw new ConfigStoreException(ConfigStoreError.OwnerLength);
                }
                
                m_owner = value;
            }
        }

        public string Thumbprint
        {
            get;
            set;
        }

        public long ID
        {
            get;
            set;
        }
        
        public byte[] Data
        {
            get
            {
                return m_data;
            }
            set
            {
                m_data = value;
            }
        }

        public DateTime CreateDate
        {
            get;
            set;
        }

        public DateTime ValidStartDate
        {
            get;
            set;
        }

        public DateTime ValidEndDate
        {
            get;
            set;
        }

        public bool ForIncoming
        {
            get;
            set;
        }

        public bool ForOutgoing
        {
            get;
            set;
        }

        public EntityStatus Status
        {
            get;
            set;
        }
        
        public bool HasData
        {
            get
            {
                return (this.Data != null && this.Data.Length > 0);
            }
        }                              

        public bool IsValid(DateTime testDate)
        {
            return (this.ValidStartDate <= testDate && testDate <= this.ValidEndDate);
        }

        public void ClearData()
        {
            m_data = null;
        }

        public void ValidateHasData()
        {
            if (!m_data.Any())
            {
                throw new ConfigStoreException(ConfigStoreError.MissingCertificateData);
            }
        }
        
        public X509Certificate2 ToX509Certificate()
        {
            return new X509Certificate2(this.Data);
        }
        
        public static X509Certificate2Collection ToX509Collection(Anchor[] anchors)
        {
            if (!anchors.Any())
            {
                return null;
            }
            
            X509Certificate2Collection x509Collection = new X509Certificate2Collection();
            if (anchors != null)
            {
                for (int i = 0; i < anchors.Length; ++i)
                {
                    x509Collection.Add(anchors[i].ToX509Certificate());
                }
            }            
            return x509Collection;
        }
    }
}